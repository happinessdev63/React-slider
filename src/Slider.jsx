'use strict';

var React = require('react');
var Tooltip = require('rc-tooltip');
var DomUtils = require('rc-util').Dom;

function pauseEvent(e) {
  if (e.stopPropagation) {
    e.stopPropagation();
  }
  if (e.preventDefault) {
    e.preventDefault();
  }
}

function prefixClsFn(prefixCls) {
  var args = Array.prototype.slice.call(arguments, 1);
  return args.map((s)=> {
    return prefixCls + '-' + s;
  }).join(' ');
}

var Slider = React.createClass({
  propTypes: {
    min: React.PropTypes.number,
    max: React.PropTypes.number,
    step: React.PropTypes.number,
    value: React.PropTypes.number,
    index: React.PropTypes.number,
    marks: React.PropTypes.array,
    isIncluded: React.PropTypes.bool,
    className: React.PropTypes.string,
    disabled: React.PropTypes.bool,
    onBeforeChange: React.PropTypes.func,
    onChange: React.PropTypes.func,
    onAfterChange: React.PropTypes.func
  },

  getDefaultProps: function() {
    return {
      min: 0,
      max: 100,
      step: 1,
      value: 0,
      marks: [],
      isIncluded: true,
      className: 'rc-slider',
      disabled: false,
      index: 0
    };
  },

  getInitialState: function() {
    var props = this.props;
    var value = this._trimAlignValue(props.value);
    var marksLen = props.marks.length;
    if (marksLen > 0) {
      value = ((props.max - props.min) / (marksLen - 1)) * (props.index);
      value = value.toFixed(5);
    }

    return {
      value: value,
      active: props.disabled ? '' : ((value > props.min || props.index > 0) ? 'active' : '')
    };
  },

  componentWillReceiveProps: function(newProps) {
    var value = newProps.value;
    this.state.value = this._trimAlignValue(value, newProps);
  },

  componentWillUnmount: function() {
    if (this._onHandleResizeListener) {
      this._onHandleResizeListener.remove();
    }
  },

  getValue: function() {
    return this.state.value;
  },

  getIndex: function() {
    var props = this.props;
    if (props.marks.length === 0) {
      return 0;
    }
    var value = this.state.value;
    var unit = ((props.max - props.min) / (props.marks.length - 1)).toFixed(5);
    return Math.floor(value / unit);
  },

  _trimAlignValue: function(val, props) {
    props = props || this.props;

    var step = props.marks.length > 0 ? (props.max - props.min) / (props.marks.length - 1) : props.step;

    if (val <= props.min) {
      val = props.min;
    }
    if (val >= props.max) {
      val = props.max;
    }

    var valModStep = (val - props.min) % step;
    var alignValue = val - valModStep;

    if (Math.abs(valModStep) * 2 >= step) {
      alignValue += (valModStep > 0) ? step : (-step);
    }

    return parseFloat(alignValue.toFixed(5));
  },

  _calcOffset: function(value) {
    var ratio = (value - this.props.min) / (this.props.max - this.props.min);
    return ratio * 100 + '%';
  },

  _calcValue: function(offset) {
    var ratio = offset / this.getSliderLength();
    return ratio * (this.props.max - this.props.min) + this.props.min;
  },

  _calValueByPos: function (position, callback) {
    var pixelOffset = position - this.getSliderStart();
    // pixelOffset -= (this.state.handleSize / 2);

    var nextValue = this._trimAlignValue(this._calcValue(pixelOffset));

    this.setState({value: nextValue, active: 'active'}, callback);
  },

  _getMousePosition: function(e) {
    return e.pageX || (e.clientX + document.documentElement.scrollLeft);
  },

  _getTouchPosition: function (e) {
    var touch = e.touches[0];
    return touch.pageX;
  },

  _triggerEvents: function(event) {
    var props = this.props;
    var hasMarks = props.marks && props.marks.length > 0;
    if (props[event]) {
      props[event](hasMarks ? this.getIndex() : this.state.value);
    }
  },

  _addEventHandles: function(type) {
    if (type === 'touch') {
      // just work for chrome iOS Safari and Android Browser
      this._onTouchMoveListener = DomUtils.addEventListener(document, 'touchmove', this._onTouchMove);
      this._onTouchUpListener = DomUtils.addEventListener(document, 'touchend', this._onTouchUp);
    }

    if (type === 'mouse') {
      this._onMouseMoveListener = DomUtils.addEventListener(document, 'mousemove', this._onMouseMove);
      this._onMouseUpListener = DomUtils.addEventListener(document, 'mouseup', this._onMouseUp);
    }
  },

  _removeEventHandles: function (type) {
    if (type === 'touch') {
      this._onTouchMoveListener.remove();
      this._onTouchUpListener.remove();
    }

    if (type === 'mouse') {
      this._onMouseMoveListener.remove();
      this._onMouseUpListener.remove();
    }
  },

  _start: function(position) {
    if (document.activeElement) {
      document.activeElement.blur();
    }

    this._triggerEvents('onBeforeChange');

    this.setState({
      startValue: this.state.value,
      startPosition: position
    });
  },

  _end: function(type) {
    this._removeEventHandles(type);
    this.setState(this._triggerEvents.bind(this, 'onAfterChange'));
  },

  _onMouseUp: function() {
    this._end('mouse');
  },

  _onTouchUp: function() {
    this._end('touch');
  },

  _onMouseMove: function(e) {
    var position = this._getMousePosition(e);
    this._handleMove(e, position);
  },

  _onTouchMove: function(e) {
    if (e.touches.length > 1 || (e.type === 'touchend' && e.touches.length > 0)) {
      return;
    }

    var position = this._getTouchPosition(e);

    this._handleMove(e, position);
  },

  _handleMove: function(e, position) {
    pauseEvent(e);
    // var position = this._getMousePosition(e);
    var props = this.props;
    var state = this.state;

    var value = state.value;
    var oldValue = value;

    var diffPosition = position - state.startPosition;

    var diffValue = diffPosition / this.getSliderLength() * (props.max - props.min);
    var newValue = this._trimAlignValue(state.startValue + diffValue);

    value = newValue;

    if (newValue !== oldValue) {
      this.setState({value: value, active: 'active'}, this._triggerEvents.bind(this, 'onChange'));
    }
  },

  getSliderLength: function() {
    var slider = this.refs.slider;
    if (!slider) {
      return 0;
    }

    return slider.getDOMNode().clientWidth;
  },

  getSliderStart: function() {
    var slider = this.refs.slider.getDOMNode();
    var rect = slider.getBoundingClientRect();

    return rect.left;
  },

  handleTouchStart: function(e) {
    if (this.props.disabled || e.touches.length > 1 || (e.type === 'touchend' && e.touches.length > 0)) {
      return;
    }

    var position = this._getTouchPosition(e);
    this.startPosition = position;
    this._start(position);
    this._addEventHandles('touch');
    pauseEvent(e);
  },

  handleMouseDown: function() {
    return (e) => {
      if (this.props.disabled) {
        return;
      }
      var position = this._getMousePosition(e);
      this._start(position);
      this._addEventHandles('mouse');
      pauseEvent(e);
    };
  },

  handleSliderMouseDown: function(e) {
    if (this.props.disabled) {
      return;
    }
    var position = this._getMousePosition(e);
    this._calValueByPos(position,
      () => {
        this._triggerEvents('onChange');
        this._start(position);
        this._addEventHandles('mouse');
      }
    );

    pauseEvent(e);
  },

  renderSteps: function() {
    var props = this.props;
    var marksLen = props.marks.length;
    var stepNum = marksLen > 0 ? marksLen : Math.floor((props.max - props.min) / props.step) + 1;
    var unit = 100 / (stepNum - 1);

    var prefixCls = props.className;
    var stepClassName = prefixClsFn(prefixCls, 'step');

    var elements = [];
    for (var i = 0; i < stepNum; i++) {
      var offset = unit * i + '%';
      var style = {
        left: offset
      };
      var className = prefixClsFn(prefixCls, 'dot');
      if (props.isIncluded) {
        if (i <= this.getIndex() || (this._calcValue(offset) <= this.getValue())) {
          className = prefixClsFn(prefixCls, 'dot', 'dot-active');
        }
      } else {
        className = (i === this.getIndex()) ? prefixClsFn(prefixCls, 'dot', 'dot-active') : className;
      }

      elements[i] = (
        <span className={className} style={style} ref={'step' + i}></span>
      );
    }

    return (
      <div className={stepClassName}>
        {elements}
      </div>
    );
  },

  renderMark: function(i) {
    var marks = this.props.marks;
    var marksLen = marks.length;
    var unit = 100 / (marksLen - 1);
    var offset = unit * i;

    var style = {
      width: unit / 2 + '%'
    };

    if (i === marksLen - 1) {
      style.right = -unit / 4 + '%';
    } else {
      style.left = i > 0 ? offset - unit / 4 + '%' : -unit / 4 + '%';
    }

    var prefixCls = this.props.className;
    var className = prefixClsFn(prefixCls, 'mark-text');

    if (this.props.isIncluded) {
      if (i <= this.getIndex()) {
        className = prefixClsFn(prefixCls, 'mark-text', 'mark-text-active');
      }
    } else {
      className = (i === this.getIndex()) ? prefixClsFn(prefixCls, 'mark-text', 'mark-text-active') : className;
    }

    return (
      <span className={className} style={style}>{this.props.marks[i]}</span>
    );
  },

  renderMarks: function() {
    var marks = this.props.marks;
    var marksLen = marks.length;
    var elements = [];
    for (var i = 0; i < marksLen; i++) {
      elements[i] = this.renderMark(i);
    }

    var prefixCls = this.props.className;
    var className = prefixClsFn(prefixCls, 'mark');

    return (
      <div className={className}>
        {elements}
      </div>
    );
  },

  renderHandle: function(offset) {
    var handleStyle = {
      left: offset
    };

    var prefixCls = this.props.className;
    var className = prefixClsFn(prefixCls, 'handle');

    if (this.state.active) {
      className =  prefixClsFn(prefixCls, 'handle', 'handle-active');
    }

    var handle =  <a className={className}
        ref = "handle"
        style = {handleStyle}
        href = "#"
        onMouseDown = {this.handleMouseDown}
        onTouchStart = {this.handleTouchStart}></a>;

    if (this.props.marks.length > 0) {
      return handle;
    } else {
      return (
        <Tooltip
          placement="top"
          overlay={<span>{this.state.value}</span>}
          delay={0}
          prefixCls={prefixClsFn(prefixCls, 'tooltip')}>
          {handle}
        </Tooltip>
      );
    }
  },

  renderTrack: function(offset) {
    var style = {
      width: offset
    };

    var prefixCls = this.props.className;
    var trackClassName = prefixClsFn(prefixCls, 'track');

    return (
      <div className={trackClassName} ref="track" style={style}></div>
    );
  },

  render: function() {
    var state = this.state;
    var props = this.props;

    var value = state.value;
    var offset = this._calcOffset(value);

    var track = this.props.isIncluded ? this.renderTrack(offset) : null;
    var handles = this.renderHandle(offset);
    var steps = (props.step > 1 || props.marks.length > 0) ? this.renderSteps() : null;
    var sliderMarks = (props.marks.length > 0) ? this.renderMarks() : null;

    var prefixCls = props.className;
    var sliderClassName = props.disabled ? prefixCls + ' ' + prefixClsFn(prefixCls, 'disabled') : prefixCls;

    return (
      <div className={sliderClassName} ref="slider" onMouseDown={this.handleSliderMouseDown}>
        {track}
        {handles}
        {steps}
        {sliderMarks}
        {this.props.children}
      </div>
    );
  }
});

module.exports = Slider;
