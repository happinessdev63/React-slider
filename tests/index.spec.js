/* eslint-disable max-len, no-undef */
import React from 'react';
import { render, mount } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';
import Slider from '..';

describe('rc-slider', () => {
  it('should render Slider with correct DOM structure', () => {
    const wrapper = render(<Slider />);
    expect(renderToJson(wrapper)).toMatchSnapshot();
  });

  it('should render Slider with default value correctly', () => {
    const wrapper = mount(<Slider defaultValue={50} />);
    expect(wrapper.state('bounds')[1]).toBe(50);
    expect(wrapper.find('.rc-slider-handle').get(0).style.cssText).toMatch(/left: 50%;/);

    const trackStyle = wrapper.find('.rc-slider-track').get(0).style.cssText;
    expect(trackStyle).toMatch(/left: 0%;/);
    expect(trackStyle).toMatch(/width: 50%;/);
    expect(trackStyle).toMatch(/visibility: visible;/);
  });

  it('should render Slider with value correctly', () => {
    const wrapper = mount(<Slider value={50} />);
    expect(wrapper.state('bounds')[1]).toBe(50);
    expect(wrapper.find('.rc-slider-handle').get(0).style.cssText).toMatch(/left: 50%;/);

    const trackStyle = wrapper.find('.rc-slider-track').get(0).style.cssText;
    expect(trackStyle).toMatch(/left: 0%;/);
    expect(trackStyle).toMatch(/width: 50%;/);
    expect(trackStyle).toMatch(/visibility: visible;/);
  });

  it('should render Range with correct DOM structure', () => {
    const wrapper = render(<Slider range />);
    expect(renderToJson(wrapper)).toMatchSnapshot();
  });

  it('should render Multi-Range with correct DOM structure', () => {
    const wrapper = render(<Slider range={3} />);
    expect(renderToJson(wrapper)).toMatchSnapshot();
  });

  it('should render Range with default value correctly', () => {
    const wrapper = mount(<Slider range defaultValue={[0, 50]} />);
    expect(wrapper.state('bounds')[0]).toBe(0);
    expect(wrapper.state('bounds')[1]).toBe(50);
    expect(wrapper.find('.rc-slider-handle').get(0).style.cssText).toMatch(/left: 0%;/);
    expect(wrapper.find('.rc-slider-handle').get(1).style.cssText).toMatch(/left: 50%;/);

    const trackStyle = wrapper.find('.rc-slider-track').get(0).style.cssText;
    expect(trackStyle).toMatch(/left: 0%;/);
    expect(trackStyle).toMatch(/width: 50%;/);
    expect(trackStyle).toMatch(/visibility: visible;/);
  });

  it('should render Multi-Range with default value correctly', () => {
    const wrapper = mount(<Slider range={3} defaultValue={[0, 25, 50, 75]} />);
    expect(wrapper.state('bounds')[0]).toBe(0);
    expect(wrapper.state('bounds')[1]).toBe(25);
    expect(wrapper.state('bounds')[2]).toBe(50);
    expect(wrapper.state('bounds')[3]).toBe(75);
    expect(wrapper.find('.rc-slider-handle').get(0).style.cssText).toMatch(/left: 0%;/);
    expect(wrapper.find('.rc-slider-handle').get(1).style.cssText).toMatch(/left: 25%;/);
    expect(wrapper.find('.rc-slider-handle').get(2).style.cssText).toMatch(/left: 50%;/);
    expect(wrapper.find('.rc-slider-handle').get(3).style.cssText).toMatch(/left: 75%;/);

    const track1Style = wrapper.find('.rc-slider-track').get(0).style.cssText;
    expect(track1Style).toMatch(/left: 0%;/);
    expect(track1Style).toMatch(/width: 25%;/);
    expect(track1Style).toMatch(/visibility: visible;/);

    const track2Style = wrapper.find('.rc-slider-track').get(1).style.cssText;
    expect(track2Style).toMatch(/left: 25%;/);
    expect(track2Style).toMatch(/width: 25%;/);
    expect(track2Style).toMatch(/visibility: visible;/);

    const track3Style = wrapper.find('.rc-slider-track').get(2).style.cssText;
    expect(track3Style).toMatch(/left: 50%;/);
    expect(track3Style).toMatch(/width: 25%;/);
    expect(track3Style).toMatch(/visibility: visible;/);
  });

  it('should render Range with value correctly', () => {
    const wrapper = mount(<Slider range value={[50, 100]} />);
    expect(wrapper.state('bounds')[0]).toBe(50);
    expect(wrapper.state('bounds')[1]).toBe(100);
    expect(wrapper.find('.rc-slider-handle').get(0).style.cssText).toMatch(/left: 50%;/);
    expect(wrapper.find('.rc-slider-handle').get(1).style.cssText).toMatch(/left: 100%;/);

    const trackStyle = wrapper.find('.rc-slider-track').get(0).style.cssText;
    expect(trackStyle).toMatch(/left: 50%;/);
    expect(trackStyle).toMatch(/width: 50%;/);
    expect(trackStyle).toMatch(/visibility: visible;/);
  });

  it('should render dots correctly when `dots=true`', () => {
    const sliderWrapper = mount(<Slider value={50} step={10} dots />);
    expect(sliderWrapper.find('.rc-slider-dot').length).toBe(11);
    expect(sliderWrapper.find('.rc-slider-dot-active').length).toBe(6);

    const rangeWrapper = mount(<Slider range value={[20, 50]} step={10} dots />);
    expect(rangeWrapper.find('.rc-slider-dot').length).toBe(11);
    expect(rangeWrapper.find('.rc-slider-dot-active').length).toBe(4);
  });

  it('should render marks correctly when `marks` is not an empty object', () => {
    const marks = { 0: '0', 30: '30', 100: '100' };

    const sliderWrapper = mount(<Slider value={30} marks={marks} />);
    expect(sliderWrapper.find('.rc-slider-mark-text').length).toBe(3);
    expect(sliderWrapper.find('.rc-slider-mark-text').get(0).innerHTML).toBe('0');
    expect(sliderWrapper.find('.rc-slider-mark-text').get(1).innerHTML).toBe('30');
    expect(sliderWrapper.find('.rc-slider-mark-text').get(2).innerHTML).toBe('100');

    const rangeWrapper = mount(<Slider range value={[0, 30]} marks={marks} />);
    expect(rangeWrapper.find('.rc-slider-mark-text').length).toBe(3);
  });

  it('should not set value greater than `max` or smaller `min`', () => {
    const sliderWithMinWrapper = mount(<Slider value={0} min={10} />);
    expect(sliderWithMinWrapper.state('bounds')[1]).toBe(10);

    const sliderWithMaxWrapper = mount(<Slider value={100} max={90} />);
    expect(sliderWithMaxWrapper.state('bounds')[1]).toBe(90);

    const rangeWrapper = mount(<Slider range value={[0, 100]} min={10} max={90} />);
    expect(rangeWrapper.state('bounds')[0]).toBe(10);
    expect(rangeWrapper.state('bounds')[1]).toBe(90);
  });

  it('should render a vertical slider, when `vertical` is true', () => {
    const wrapper = mount(<Slider vertical />);
    expect(wrapper.find('.rc-slider-vertical').length).toBe(1);
  });

  it('should not call onChange when value is the same', () => {
    const values = [];
    const handler = (e) => {
      values.push(e);
    };

    const wrapper = mount(<Slider onChange={handler}/>);
    const handleWrapper = wrapper.find('.rc-slider-handle');
    handleWrapper.simulate('mousedown');
    handleWrapper.simulate('mousemove');
    handleWrapper.simulate('mouseup');

    expect(values.length).toBe(0);
  });

  it('should set `dragOffset` to correct value when the left handle is clicked off-center', () => {
    const wrapper = mount(<Slider />);
    wrapper.node.refs.slider.clientWidth = 100; // jsdom doesn't provide clientWidth
    const leftHandle = wrapper.find('.rc-slider-handle').get(0);
    wrapper.simulate('mousedown', {
      type: 'mousedown',
      target: leftHandle,
      pageX: 5, button: 0,
      stopPropagation() {},
      preventDefault() {},
    });
    expect(wrapper.node.dragOffset).toBe(5);
  });

  it('should respect `dragOffset` while dragging the handle via MouseEvents', () => {
    const wrapper = mount(<Slider />);
    wrapper.node.refs.slider.clientWidth = 100; // jsdom doesn't provide clientWidth
    const leftHandle = wrapper.find('.rc-slider-handle').get(0);
    wrapper.simulate('mousedown', {
      type: 'mousedown',
      target: leftHandle,
      pageX: 5, button: 0,
      stopPropagation() {},
      preventDefault() {},
    });
    expect(wrapper.node.dragOffset).toBe(5);
    wrapper.node.onMouseMove({ // to propagation
      type: 'mousemove',
      target: leftHandle,
      pageX: 14, button: 0,
      stopPropagation() {},
      preventDefault() {},
    });
    expect(wrapper.node.getValue()).toBe(9);
  });

  it('should set `dragOffset` to 0 when the MouseEvent target isn\'t a handle', () => {
    const wrapper = mount(<Slider />);
    wrapper.node.refs.slider.clientWidth = 100; // jsdom doesn't provide clientWidth
    const sliderTrack = wrapper.find('.rc-slider-track').get(0);
    wrapper.simulate('mousedown', {
      type: 'mousedown',
      target: sliderTrack,
      pageX: 5, button: 0,
      stopPropagation() {},
      preventDefault() {},
    });
    expect(wrapper.node.dragOffset).toBe(0);
  });

  it('should set `dragOffset` to correct value when the left handle is touched off-center', () => {
    const wrapper = mount(<Slider />);
    wrapper.node.refs.slider.clientWidth = 100; // jsdom doesn't provide clientWidth
    const leftHandle = wrapper.find('.rc-slider-handle').get(0);
    wrapper.simulate('touchstart', {
      type: 'touchstart',
      target: leftHandle,
      touches: [{ pageX: 5 }],
      stopPropagation() {},
      preventDefault() {},
    });
    expect(wrapper.node.dragOffset).toBe(5);
  });

  it('should respect `dragOffset` while dragging the handle via TouchEvents', () => {
    const wrapper = mount(<Slider />);
    wrapper.node.refs.slider.clientWidth = 100; // jsdom doesn't provide clientWidth
    const leftHandle = wrapper.find('.rc-slider-handle').get(0);
    wrapper.simulate('touchstart', {
      type: 'touchstart',
      target: leftHandle,
      touches: [{ pageX: 5 }],
      stopPropagation() {},
      preventDefault() {},
    });
    expect(wrapper.node.dragOffset).toBe(5);
    wrapper.node.onTouchMove({ // to propagation
      type: 'touchmove',
      target: leftHandle,
      touches: [{ pageX: 14 }],
      stopPropagation() {},
      preventDefault() {},
    });
    expect(wrapper.node.getValue()).toBe(9);
  });

  it('should set `dragOffset` to 0 when the TouchEvent target isn\'t a handle', () => {
    const wrapper = mount(<Slider />);
    wrapper.node.refs.slider.clientWidth = 100; // jsdom doesn't provide clientWidth
    const sliderTrack = wrapper.find('.rc-slider-track').get(0);
    wrapper.simulate('touchstart', {
      type: 'touchstart',
      target: sliderTrack,
      touches: [{ pageX: 5 }],
      stopPropagation() {},
      preventDefault() {},
    });
    expect(wrapper.node.dragOffset).toBe(0);
  });
});
