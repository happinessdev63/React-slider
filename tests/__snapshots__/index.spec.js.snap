exports[`rc-slider should render Multi-Range with correct DOM structure 1`] = `
<div
  class="rc-slider">
  <div
    class="rc-slider-rail" />
  <div
    class="rc-slider-track rc-slider-track-1"
    style="visibility:visible;left:0%;width:0%;" />
  <div
    class="rc-slider-track rc-slider-track-2"
    style="visibility:visible;left:0%;width:0%;" />
  <div
    class="rc-slider-track rc-slider-track-3"
    style="visibility:visible;left:0%;width:0%;" />
  <div
    class="rc-slider-step" />
  <div
    class="rc-slider-handle rc-slider-handle-1 rc-slider-handle-lower"
    style="left:0%;" />
  <div
    class="rc-slider-handle rc-slider-handle-2"
    style="left:0%;" />
  <div
    class="rc-slider-handle rc-slider-handle-3"
    style="left:0%;" />
  <div
    class="rc-slider-handle rc-slider-handle-4 rc-slider-handle-upper"
    style="left:0%;" />
  <div
    class="rc-slider-mark" />
</div>
`;

exports[`rc-slider should render Range with correct DOM structure 1`] = `
<div
  class="rc-slider">
  <div
    class="rc-slider-rail" />
  <div
    class="rc-slider-track rc-slider-track-1"
    style="visibility:visible;left:0%;width:0%;" />
  <div
    class="rc-slider-step" />
  <div
    class="rc-slider-handle rc-slider-handle-1 rc-slider-handle-lower"
    style="left:0%;" />
  <div
    class="rc-slider-handle rc-slider-handle-2 rc-slider-handle-upper"
    style="left:0%;" />
  <div
    class="rc-slider-mark" />
</div>
`;

exports[`rc-slider should render Slider with correct DOM structure 1`] = `
<div
  class="rc-slider">
  <div
    class="rc-slider-rail" />
  <div
    class="rc-slider-track rc-slider-track-1"
    style="visibility:visible;left:0%;width:0%;" />
  <div
    class="rc-slider-step" />
  <div
    class="rc-slider-handle rc-slider-handle-2 rc-slider-handle-upper"
    style="left:0%;" />
  <div
    class="rc-slider-mark" />
</div>
`;
